### PR TITLE
Add Cropping Options to Eng Diff Calib Tab

### LINKED ISSUE #26849: Add ability to crop new or existing calibration in ISIS Engineering GUI
As a user I would like the option to crop any existing or new calibration. The functionality should give the option to select by bank of the selected instrument or to input a comma-separated list of spectrum numbers.

### PR DESCRIPTION
**Pull request by cailafinn**
**Description of work.**
Cropping can now be done when creating a new calibration.
The GUI now allows for cropping by bank (1 or 2), or by specific spectrum numbers. 

**To test:**
1. Open Engineering Diffraction 2 GUI
2. Make sure that cropping options are greyed out when Load is selected.
3. Select Create New again (Valid Options: V: 307521 CeO2: 305738)
4. Check that the combo box selector appears when the cropping check box is selected 
5. Check selecting custom spectra reveals a Text entry
6. Check that only sane values can be entered and error messages make sense. (1200-1400 should guarantee an output)
7. Check that outputs are restricted by the cropping and have sane names. 

Fixes #26849 

*This does not require release notes* because **the GUI is not user ready.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+        # Connect view signals to local functions
+        self.view.set_on_combo_changed(self.on_combo_changed)
+        self.view.set_on_custom_spectra_changed(self.on_spectra_changed)
+
+        self.on_combo_changed(0)
+
+    # Signal Activated Functions
+
+    def on_combo_changed(self, index_or_string_i_dunno):  # TODO: Does this signal send and index or text?
```
Comment (Harrietbrown) [inline scripts/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py]: This TODO should be sorted out in this PR

**Review by Harrietbrown — CHANGES_REQUESTED** [review]
Is there any way you could add extra validation to the cropping values allowed. Is there any way to say what values are valid before running the calibration?

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py [hunk 2]
```diff
@@ -185,35 +202,82 @@ def _plot_difc_tzero():
...
         fig.show()
 
-    def run_calibration(self, sample_ws, van_integration, van_curves, full_calib_ws=None):
+    @staticmethod
+    def _plot_difc_tzero_single_bank_or_custom(bank, custom):
+        if not custom:
+            bank_ws = Ads.retrieve("engggui_difc_zero_peaks_bank_" + str(bank))
+        else:
+            bank_ws = Ads.retrieve("engggui_difc_zero_peaks_bank_cropped")
+            bank = "Cropped"
```
[the remark this answers is not in the document]
Comment (ConorMFinn) [reply scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py]: No, there isn't good spot.

**Review by Harrietbrown — CHANGES_REQUESTED** [review]
Good job tidying up those functions, I think it would be tricky to make it any tidyer.

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py [hunk 3]
```diff
@@ -231,49 +240,35 @@ def run_calibration(self, sample_ws, van_integration, van_curves, bank, spectrum
...
-                                     VanCurvesWorkspace=van_curves,
-                                     Bank=calib_bank,
-                                     FittedPeaks=table_name)
+        kwargs = {
+            "InputWorkspace": sample_ws,
+            "VanIntegrationWorkspace": van_integration,
+            "VanCurvesWorkspace": van_curves
+        }
+
+        def run_engg_calibrate(kwargs_to_pass):
```
[the remark this answers is not in the document]
Comment (ConorMFinn) [reply scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py]: This is my personal preference, but I like having the `**` operator refactored out of the main body of the function.

File: scripts/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py [hunk 4]
```diff
@@ -64,7 +64,7 @@ def is_valid(self):
 
     def set_custom_visibility(self, visible):
         if visible:
-            self.custom_valid = False
+            self.on_spectra_changed(self.view.get_custom_spectra_text())
```
[the remark this answers is not in the document]
Comment (ConorMFinn) [reply scripts/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py]: Fixed now

File: scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Harrietbrown) [reply scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py]: Fair enough

**Comment by Harrietbrown** [comment]
I am guessing that there isn't a way to premptivly validate the cropping range for a given input?

**Comment by ConorMFinn** [comment]
> I am guessing that there isn't a way to premptivly validate the cropping range for a given input?

I haven't been able to do this, but it should be a little less picky now about ranges.

**Review by Harrietbrown — APPROVED** [review]
Good job. The cropping GUI is very simple to use and understand.

**Pull request merged**